File: scripts/download_logos.py
#!/usr/bin/env python3
"""Download the first image from a search query and save it in a named folder.

Usage:
    python scripts/download_logos.py --input list.txt
    python scripts/download_logos.py --query OpenAI --query Google
    python scripts/download_logos.py --input list.txt --output downloaded_images
"""

import argparse
import os
import re
import time
import urllib.parse
import urllib.request
from pathlib import Path

HEADERS = {
    'User-Agent': (
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
        'AppleWebKit/537.36 (KHTML, like Gecko) '
        'Chrome/124.0.0.0 Safari/537.36'
    ),
    'Accept-Language': 'en-US,en;q=0.9',
}

SEARCH_URL = 'https://www.bing.com/images/search?q={query}'

INVALID_PATH_CHARS = r'<>:"/\\|?*'


def sanitize_name(name: str) -> str:
    name = name.strip()
    name = re.sub(r'\s+', ' ', name)
    return ''.join('_' if c in INVALID_PATH_CHARS else c for c in name)


def fetch_search_html(query: str, timeout: int = 15) -> str | None:
    url = SEARCH_URL.format(query=urllib.parse.quote_plus(query))
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req, timeout=timeout) as response:
        charset = response.headers.get_content_charset('utf-8')
        return response.read().decode(charset, errors='ignore')


def extract_first_image_url(html: str) -> str | None:
    # Try the JSON field used by Bing image results pages
    match = re.search(r'"murl"\s*:\s*"(https?://[^"\\]+)"', html)
    if match:
        return match.group(1).replace('\\u002f', '/')

    # Fallback to the first <img class="mimg" ... src="...">
    match = re.search(r'<img[^>]+class="[^"]*mimg[^"]*"[^>]+src="(https?://[^"]+)"', html)
    if match:
        return match.group(1)

    # Fallback for data-src attributes
    match = re.search(r'<img[^>]+data-src="(https?://[^"]+)"', html)
    if match:
        return match.group(1)

    return None


def download_image(url: str, dest_path: Path, timeout: int = 30) -> bool:
    req = urllib.request.Request(url, headers=HEADERS)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as response:
            data = response.read()
            dest_path.write_bytes(data)
            return True
    except Exception as exc:
        print(f'  [error] failed to download {url}: {exc}')
        return False


def file_extension_from_url(url: str) -> str:
    parsed = urllib.parse.urlparse(url)
    root, ext = os.path.splitext(parsed.path)
    if ext and re.fullmatch(r'\.(jpg|jpeg|png|gif|webp|svg|bmp|ico|jfif|tiff?)', ext, re.IGNORECASE):
        return ext
    return '.jpg'


def load_queries_from_file(path: Path) -> list[str]:
    if not path.exists():
        raise FileNotFoundError(f'Input file not found: {path}')
    with path.open('r', encoding='utf-8') as stream:
        lines = [line.strip() for line in stream if line.strip() and not line.startswith('#')]
    return lines


def main() -> None:
    parser = argparse.ArgumentParser(description='Download the first image result for each search term.')
    parser.add_argument('--input', '-i', type=Path, default=Path('list.txt'),
                        help='Text file with one search query per line.')
    parser.add_argument('--output', '-o', type=Path, default=Path('downloaded_images'),
                        help='Output directory for downloaded company folders.')
    parser.add_argument('--query', '-q', action='append', default=[],
                        help='Add a single search query. Can be repeated.')
    parser.add_argument('--limit', '-l', type=int, default=0,
                        help='Limit the number of queries downloaded from the input list. 0 means no limit.')
    parser.add_argument('--delay', '-d', type=float, default=1.0,
                        help='Seconds to wait between search requests.')
    args = parser.parse_args()

    queries: list[str] = []
    if args.query:
        queries.extend(args.query)

    if args.input and args.input.exists():
        queries.extend(load_queries_from_file(args.input))

    queries = [q for q in queries if q.strip()]
    if not queries:
        parser.error('No search queries provided. Use --query or supply an input file.')

    if args.limit > 0:
        queries = queries[: args.limit]

    args.output.mkdir(parents=True, exist_ok=True)

    print(f'Downloading first image for {len(queries)} queries into: {args.output}\n')
    for index, query in enumerate(queries, start=1):
        safe_name = sanitize_name(query)
        folder = args.output / safe_name
        folder.mkdir(parents=True, exist_ok=True)

        print(f'[{index}/{len(queries)}] {query}')
        html = fetch_search_html(query)
        if html is None:
            print('  [error] unable to fetch search result HTML')
            continue

        image_url = extract_first_image_url(html)
        if not image_url:
            print('  [error] no image URL found in search result HTML')
            continue

        ext = file_extension_from_url(image_url)
        dest_file = folder / f'logo{ext}'
        success = download_image(image_url, dest_file)
        if success:
            print(f'  saved: {dest_file}')
        time.sleep(args.delay)


if __name__ == '__main__':
    main()

File: scripts/download_logos.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

const INVALID_PATH_CHARS = /[<>:"/\\|?*]/g;
const BING_IMAGE_SEARCH = 'https://www.bing.com/images/search?q=';
const HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  'Accept-Language': 'en-US,en;q=0.9',
};

function sanitizeName(name) {
  return name
    .trim()
    .replace(/\s+/g, ' ')
    .replace(INVALID_PATH_CHARS, '_');
}

function parseArgs(argv) {
  const args = { input: 'list.txt', output: 'downloaded_images', query: [], limit: 0, delay: 1.0 };
  const tokens = argv.slice(2);
  for (let i = 0; i < tokens.length; i += 1) {
    const token = tokens[i];
    if (token === '--input' || token === '-i') {
      args.input = tokens[++i] || args.input;
    } else if (token === '--output' || token === '-o') {
      args.output = tokens[++i] || args.output;
    } else if (token === '--query' || token === '-q') {
      const value = tokens[++i];
      if (value) args.query.push(value);
    } else if (token === '--limit' || token === '-l') {
      args.limit = Number(tokens[++i]) || 0;
    } else if (token === '--delay' || token === '-d') {
      args.delay = Number(tokens[++i]) || 1.0;
    } else if (token === '--help' || token === '-h') {
      printUsageAndExit();
    }
  }
  return args;
}

function printUsageAndExit() {
  const text = `Usage: node scripts/download_logos.js [options]

Options:
  --input, -i   Path to query file (default: list.txt)
  --output, -o  Output directory (default: downloaded_images)
  --query, -q   Search query; repeatable
  --limit, -l   Maximum number of queries to process
  --delay, -d   Seconds to wait between requests (default: 1.0)
  --help, -h    Show this message
`;
  process.stdout.write(text);
  process.exit(0);
}

async function fetchHtml(query) {
  const url = BING_IMAGE_SEARCH + encodeURIComponent(query);
  const response = await fetch(url, { headers: HEADERS });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status} ${response.statusText}`);
  }
  return response.text();
}

function extractFirstImageUrl(html) {
  const murl = html.match(/"murl"\s*:\s*"(https?:\/\/[^"\\]+)"/i);
  if (murl) {
    return murl[1].replace(/\\u002f/g, '/');
  }

  const img = html.match(/<img[^>]+class="[^"]*mimg[^"]*"[^>]+src="(https?:\/\/[^"\\]+)"/i);
  if (img) {
    return img[1];
  }

  const dataSrc = html.match(/<img[^>]+data-src="(https?:\/\/[^"\\]+)"/i);
  if (dataSrc) {
    return dataSrc[1];
  }

  return null;
}

function fileExtensionFromUrl(imageUrl) {
  try {
    const parsed = new URL(imageUrl);
    const ext = path.extname(parsed.pathname).toLowerCase();
    if (/\.(jpg|jpeg|png|gif|webp|svg|bmp|ico|jfif|tiff?)$/.test(ext)) {
      return ext;
    }
  } catch (err) {
    // ignore invalid URL
  }
  return '.jpg';
}

function loadQueriesFromFile(filePath) {
  const absolutePath = path.resolve(process.cwd(), filePath);
  if (!fs.existsSync(absolutePath)) {
    throw new Error(`Input file not found: ${absolutePath}`);
  }
  return fs.readFileSync(absolutePath, 'utf-8')
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith('#'));
}

async function downloadImage(imageUrl, destination) {
  const response = await fetch(imageUrl, { headers: HEADERS });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status} ${response.statusText}`);
  }
  const buffer = await response.arrayBuffer();
  fs.writeFileSync(destination, Buffer.from(buffer));
}

async function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

async function run() {
  const options = parseArgs(process.argv);
  const queries = [];

  if (options.query.length > 0) {
    queries.push(...options.query);
  }
  if (options.input) {
    queries.push(...loadQueriesFromFile(options.input));
  }

  const uniqueQueries = [...new Set(queries.map((q) => q.trim()).filter(Boolean))];
  if (uniqueQueries.length === 0) {
    printUsageAndExit();
  }

  const selectedQueries = options.limit > 0 ? uniqueQueries.slice(0, options.limit) : uniqueQueries;
  const outDir = path.resolve(process.cwd(), options.output);
  fs.mkdirSync(outDir, { recursive: true });

  console.log(`Downloading first image for ${selectedQueries.length} queries into: ${outDir}\n`);

  for (let i = 0; i < selectedQueries.length; i += 1) {
    const query = selectedQueries[i];
    const safeName = sanitizeName(query);
    const folder = path.join(outDir, safeName);
    fs.mkdirSync(folder, { recursive: true });
    console.log(`[${i + 1}/${selectedQueries.length}] ${query}`);

    try {
      const html = await fetchHtml(query);
      const imageUrl = extractFirstImageUrl(html);
      if (!imageUrl) {
        console.error('  [error] no image URL found');
        continue;
      }
      const ext = fileExtensionFromUrl(imageUrl);
      const destination = path.join(folder, `logo${ext}`);
      await downloadImage(imageUrl, destination);
      console.log(`  saved: ${destination}`);
    } catch (error) {
      console.error(`  [error] ${error.message}`);
    }

    if (i < selectedQueries.length - 1) {
      await sleep(options.delay);
    }
  }
}

run().catch((error) => {
  console.error(error.message);
  process.exit(1);
});
